synapse/lib/base.py

```diff
@@ -0,0 +1,571 @@
+            except Exception as e:
+                logger.exception('base %s error with mesg %s', self, mesg)
+
+        return ret
+
+    async def fini(self):
+        '''
+        Shut down the object and notify any onfini() coroutines.
+
+        Returns:
+            Remaining ref count
+        '''
+        if self.isfini:
+            return
+
+        self._syn_refs -= 1
+        if self._syn_refs > 0:
+            return self._syn_refs
+
+        self.isfini = True
+
+        fevt = self.finievt
+        for fini in self._fini_funcs:
+
+            try:
+                rv = fini()
+                if asyncio.iscoroutine(rv):
+                    await rv
+
+            except Exception as e:
+                logger.exception('fini failed')
+
+        if fevt is not None:
+            fevt.set()
+        self._syn_funcs.clear()
+        del self._fini_funcs[:]
+        return 0
+
+    async def waitfini(self, timeout=None):
+        '''
+        Wait for the event bus to fini()
+
+        Example:
+
+            bus.waitfini(timeout=30)
+
+        '''
+
+        if self.isfini:
+            return True
+
+        if self.finievt is None:
+            self.finievt = asyncio.Event(loop=self.loop)
+
+        try:
+            await asyncio.wait_for(self.finievt.wait(), timeout, loop=self.loop)
+        except asyncio.TimeoutError:
+            return None
+        return True
+
+    def main(self):
+        '''
+        Helper function to block until shutdown ( and handle ctrl-c and SIGTERM).
+
+        Examples:
+            Run a event bus, wait until main() has returned, then do other stuff::
+
+                foo = EventBus()
+                foo.main()
+                dostuff()
+
+        Notes:
+            This does fire a 'ebus:main' event prior to entering the
+            waitfini() loop.
+
+        Returns:
+            None
+        '''
+        assert self.loop == s_glob.plex.loop
+
+        doneevent = threading.Event()
+        self.onfini(doneevent.set)
+
+        async def sighandler():
+            print('Caught SIGTERM, shutting down')
+            await self.fini()
+
+        def handler():
+            asyncio.run_coroutine_threadsafe(sighandler(), loop=self.loop)
+
+        try:
+            s_glob.plex.loop.add_signal_handler(signal.SIGTERM, handler)
+        except Exception as e:  # pragma: no cover
+            logger.exception('Unable to register SIGTERM handler.')
+
+        async def asyncmain():
+            await self.fire('ebus:main')
+
+        asyncio.run_coroutine_threadsafe(asyncmain(), loop=self.loop)
+
+        try:
+            doneevent.wait()
+
+        except KeyboardInterrupt as e:
+            print('ctrl-c caught: shutting down')
+
+        finally:
+            # Avoid https://bugs.python.org/issue34680 by removing handler before closing down
+            s_glob.plex.loop.remove_signal_handler(signal.SIGTERM)
+
+            asyncio.run_coroutine_threadsafe(self.fini(), loop=self.loop)
```

Comment:
This is going to return a future (which we don't snag) and then main() will exist right away.  This could allow for interpreter shutdown prior to the fini() running.  Can we wait for this future to run?  We definitely want to give objects the chance to gracefully shut down.

Reply:
Actually, I'm not sure we should be fini'ing at all. The caller initialized the daemon, he should probably fini it too.

Reply:
And the only current caller of `main` calls main inside a context manager, so I'm gonna delete.